### DISCUSSION TITLE
Inode Value 34 for ZFS Dataset Root Node

### DISCUSSION THREAD

**Discussion by nervtech** [openzfs/zfs#15192] [General]
Hello ZFS community,

I've been exploring the intricacies of ZFS and have come across an interesting observation. It seems that for newly created datasets, the inode value for the root node is consistently set to 34. I'm curious to understand the rationale behind this specific value and how it is determined during dataset creation.

**Comment by rincebrain** [comment]
Not always.

```
$ stat -c '%i' /monolith/
4
$ stat -c '%i' /turbopool/
34
```

```
Dataset monolith [ZPL], ID 21, cr_txg 1, 228K, 34 objects

    ZIL header: claim_txg 0, claim_blk_seq 0, claim_lr_seq 0 replay_seq 0, flags 0x0


    Deadlist: 0 (0/0 comp), 88 entries

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         0    7    16K    16K   104K     512   592K    2.87  DMU dnode
        -1    1    16K    512      0     512    512  100.00  ZFS user/group/project used
        -2    1    16K    512      0     512    512  100.00  ZFS user/group/project used
         1    1    16K    512     8K     512    512  100.00  ZFS master node
         2    1    16K    512      0     512    512  100.00  SA master node
         3    1    16K    512      0     512    512  100.00  ZFS delete queue
         4    1    16K     2K     8K     512     2K  100.00  ZFS directory
```

```
$ sudo zdb -ddd turbopool/ | less
Dataset turbopool [ZPL], ID 54, cr_txg 1, 24K, 8 objects

    Deadlist: 0 (0/0 comp), 70 entries

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         0    6   128K    16K    13K     512    32K   12.50  DMU dnode
        -1    1   128K    512      0     512    512  100.00  ZFS user/group/project used
        -2    1   128K    512      0     512    512  100.00  ZFS user/group/project used
        -3    1   128K    512      0     512    512  100.00  ZFS user/group/project used
         1    1   128K    512     1K     512    512  100.00  ZFS master node
         2    1   128K    512      0     512    512  100.00  ZFS directory
         3    1   128K    512      0     512    512  100.00  ZFS directory
        32    1   128K    512      0     512    512  100.00  SA master node
        33    1   128K    512      0     512    512  100.00  ZFS delete queue
        34    1   128K    512      0     512    512  100.00  ZFS directory
```

I would assume, without having the early pool creation history in front of me, that you wind up with creating a bunch of objects now for various features that are newer that get turned over again before making the / directory object.

But it's not a hardcoded constant, AFAICT.

**Comment by lundman** [comment]
It starts on a low number, then each objected created, it gets incremented. A few are created when you create the pool and root dataset, so it will probably appear stable, at least until some code changes come along.
If anyone cares, the root inode has to be a certain value on macOS, so we need to send that instead of the zfs root inode, additionally, the first 16 are reserved, so we had to add some code to translate around them.

**Reply by nervtech** [reply]
I'm planning to store the entire dataset's file metadata in a tree-like structure, so I'm quite concerned about the inode of the root node. It seems like a safe bet to query and save the inode of the root node after creating the dataset.
